docs/README.md

```diff
@@ -0,0 +1,327 @@
+More recent proposals are located in the
+[swift-evolution][] repository.
+
+[swift-evolution]: https://github.com/apple/swift-evolution
+
+### Surveys
+
+- [ErrorHandlingRationale.rst](/docs/ErrorHandlingRationale.rst):
+  Surveys error-handling in a variety of languages,
+  and describes the rationale behind the design of error handling in Swift.
+- [weak.rst](/docs/weak.rst):
+  Discusses weak references, including the designs in different languages,
+  and proposes changes to Swift (pre-1.0).
+  <!-- NOTE: Outdated -->
+
+### Archive
+
+These documents are known to be out-of-date and are superseded by other
+documentation, primarily [The Swift Programming Language][].
+They are preserved mostly for historical interest.
+
+- [AccessControl.rst](/docs/AccessControl.swift)
+- [Arrays.rst](/docs/Arrays.rst)
+  <!-- Has additional notes on bridging that may be of general interest? -->
+- [Generics.rst](/docs/Generics.rst)
+- [ErrorHandling.rst](/docs/ErrorHandling.rst)
+- [StringDesign.rst](/docs/StringDesign.rst)
+- [TextFormatting.rst](/docs/TextFormatting.rst)
+
+[The Swift Programming Language]: https://docs.swift.org/swift-book
+
+## External Resources
+
+The official [Swift blog](https://swift.org/blog/) contains a lot of useful information,
+such as how library evolution works
+and how the compiler's new diagnostic architecture is structured,
+helping us provide more precise diagnostics.
+
+TODO: Add a new document ExternalResources.md.
+
+## Uncategorized
+
+### Needs refactoring
+
+The documents in this section might be worth breaking up
+into several documents, and linking one document from the other.
+Breaking up into components will provide greater clarity
+to contributors wanting to add new documentation.
+
+- [ARCOptimization.rst](/docs/ARCOptimization.rst):
+  Covers how ARC optimization works, with several examples.
+  TODO: Not clear if this is intended to be a topic guide
+  or a reference guide.
+- [CompilerPerformance.md](/docs/CompilerPerformance.md):
+  Thoroughly discusses different ways of measuring compiler performance
+  and common pitfalls.
+  TODO: Consider breaking up into one high-level
+  topic guide explaining key concepts and individual how-to guides that
+  can be expanded independently.
+  Also, it's not the right place to explain details of different compiler modes.
+- [DevelopmentTips.md](/docs/DevelopmentTips.md):
+  Contains an assortment of tips for better productivity when working on the
+  compiler.
+  TODO: Might be worthwhile to make a dedicated how-to guide or topic guide.
+  It might also be valuable to introduce the tips in context, and have the
+  topic guide link to all the different tips.
+- [Diagnostics.md](/docs/Diagnostics.md):
+  Describes how to write diagnostic messages and associated educational notes.
+  TODO: Consider splitting into how-tos (e.g. for educational notes)
+  and recommended practices (on writing diagnostics).
```

Comment:
It sounds like this is much more an in-depth topic than How-To; it's not ever going to be a "common task" to write an educational note...

Reply:
I kinda' used fewer words than I should have; no wonder that bites me here. 😄 

IMO the educational notes are a part of the diagnostics, so the 'recommended practices' part on diagnostics would have a section on what kind of verbiage should be used for educational notes too.

So it would probably have been more precise to write:

>   TODO: Consider splitting into how-tos (e.g. for educational notes)
>  and recommended practices (maybe one for writing diagnostics,
>  and educational notes).

---

> it's not ever going to be a "common task" to write an educational note...

I hope that's not the case 😛! My understanding (@owenv can correct me if I'm wrong) is that eventual hope is that lots of our diagnostics have educational notes explaining things in greater detail. This will also be a good way for people to get their feet wet with contributing too. You can add a good explanation of an error you found confusing but it didn't have an educational note explaining the details!

Reply:
> My understanding (@owenv can correct me if I'm wrong) is that eventual hope is that lots of our diagnostics have educational notes explaining things in greater detail.

That's the goal, but it's a _long_ ways away :)

I think the bulk of this doc can be categorized as recommended practices. I agree it would be nice to break some of it out into a how-to guide though. Diagnostic fixes are a really common kind of starter bug because many of them are relatively self-contained changes. At the same time, the relationship between all the different moving parts (.def files, format strings, etc.) can be hard to grasp at first, and a focused tutorial would be great.

Reply:
Updated the TODO here to clarify what needs to be done.